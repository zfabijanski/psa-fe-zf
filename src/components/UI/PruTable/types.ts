import { ReactNode } from "react";

export type SortOrder = "descend" | "ascend";
export interface IColumn<T> {
  field: keyof T;
  title: string;
  width?: string;
  render?: (record: T) => ReactNode;
  sorter?: (a: T, b: T) => number;
}

export interface IColumnSorting<T> {
  field: keyof T;
  order: SortOrder;
}
