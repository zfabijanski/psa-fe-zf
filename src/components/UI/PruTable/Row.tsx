import React, { Component } from "react";
import styled from "styled-components";
import { textEllipsis } from "../../../theme/textEllipsis";
import { Icon } from "../Icon";
import { IColumn, IColumnSorting } from "./types";

interface ITr {
  isSelectable?: boolean;
}

const Tr = styled.tr<ITr>`
  cursor: ${({ isSelectable }) => (isSelectable ? "pointer" : "auto")};
  border: 1px solid ${({ theme }) => theme.newColors.gray40};

  &:hover > * {
    background-color: ${({ theme }) => theme.newColors.gray5};
  }
`;

const Td = styled.td<{ isActive?: boolean; alignRight?: boolean }>`
  padding: 13px 24px 13px 12px;
  color: ${({ theme, isActive }) =>
    isActive ? theme.colors.textPrimary : theme.newColors.neutral100Default};
  font-weight: ${({ isActive }) => isActive && "600"};
  text-align: ${({ alignRight }) => alignRight && "right"};
  ${textEllipsis}
`;

const TdRight = styled(Td)`
  text-align: left;
  cursor: pointer;
`;

const IconWrapper = styled.div`
  background-color: ${({ theme }) => theme.newColors.white100};
  height: 28px;
  width: 28px;
  display: flex;
  justify-content: center;
  border: 2px solid ${({ theme }) => theme.newColors.white100};

  & path {
    stroke-width: 2px;
  }

  &:hover {
    border-color: ${({ theme }) => theme.newColors.primary100};
    background-color: ${({ theme }) => theme.newColors.primary05};

    & path {
      stroke: ${({ theme }) => theme.newColors.primary100};
    }
  }

  &:disabled {
    border-color: ${({ theme }) => theme.newColors.gray30};
    background-color: ${({ theme }) => theme.newColors.gray20};
  }
`;

interface IRowProps<T> {
  primaryKey: keyof T;
  dataRow: T;
  columns: Array<IColumn<T>>;
  sorting: IColumnSorting<T> | undefined;
  onSelect: () => void;
  onDelete: () => void;
  isSelectable?: boolean;
}

class Row<T> extends Component<IRowProps<T>> {
  private isSelectable = () =>
    this.props.isSelectable !== undefined ? this.props.isSelectable : true;

  private getAlignRight = (column: IColumn<T>) =>
    ["createDate", "lastAccessDate"].includes(column.field.toString());

  private getOrder = (column: IColumn<T>) => {
    const { sorting } = this.props;
    return sorting && sorting.field === column.field
      ? sorting.order
      : undefined;
  };

  private handleDelete = (event: React.MouseEvent) => {
    event.stopPropagation();
    this.props.onDelete();
  };

  private handleSelect = () => {
    if (this.isSelectable()) {
      this.props.onSelect();
    }
  };

  public render() {
    const { dataRow, columns } = this.props;

    const renderCell = (row: T, column: IColumn<T>) => {
      if (column.render) {
        return column.render(row);
      } else {
        return <>{row[column.field] || "-"}</>;
      }
    };

    return (
      <Tr
        key={String(dataRow[this.props.primaryKey])}
        onClick={this.handleSelect}
        isSelectable={this.isSelectable()}
      >
        {columns.map((column, index) => (
          <Td
            key={`${dataRow[this.props.primaryKey]}-${index}`}
            isActive={!!this.getOrder(column)}
            alignRight={this.getAlignRight(column)}
          >
            {renderCell(dataRow, column)}
          </Td>
        ))}
        <TdRight onClick={this.handleDelete} data-testid="delete-button">
          <IconWrapper>
            <Icon name="trash" width={16} />
          </IconWrapper>
        </TdRight>
      </Tr>
    );
  }
}

export default Row;
