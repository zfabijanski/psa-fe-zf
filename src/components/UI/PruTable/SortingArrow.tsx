import styled from "styled-components";
import { SortOrder } from "./types";
interface IArrowProps {
  order?: SortOrder;
}

const Arrow = styled.span<IArrowProps>`
  height: 25px;
  width: 25px;
  display: block;
  position: relative;
  cursor: pointer;

  &:before,
  &:after {
    border: 4px solid transparent;
    content: "";
    display: block;
    height: 0;
    right: 12px;
    top: 50%;
    position: absolute;
    width: 0;
  }
  &:before {
    border-bottom-color: ${({ order, theme }) =>
      order === "ascend" ? theme.colors.textPrimary : theme.newColors.mono40};
    margin-top: -9px;
  }
  &:after {
    border-top-color: ${({ order, theme }) =>
      order === "descend" ? theme.colors.textPrimary : theme.newColors.mono40};
    margin-top: 1px;
  }
  &:hover {
    opacity: ${({ theme }) => theme.opacityHover};
  }
`;

export const SortingArrow = ({ order }: IArrowProps) => <Arrow order={order} />;

export default SortingArrow;
