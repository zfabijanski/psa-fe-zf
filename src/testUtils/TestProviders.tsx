import { render, RenderOptions } from "@testing-library/react";
import { FC, PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { ThemeProvider } from "styled-components";
import { Normalize } from "styled-normalize";
import GlobalStyles from "theme/globalStyles";
import { theme } from "theme/theme";
import rootStore, { configureStore, rootReducer as reducer } from "AppStore";
import { BrowserRouter } from "react-router-dom";
import { Modal } from "components/UI/Modal";
import { IntlProvider } from "react-intl";
import { plMessages } from "translations/pl";
import I18nProvider from "I18nProvider";

interface TestProvidersProps {
  store?: typeof rootStore;
  renderModal?: boolean;
  realIntlProvider?: boolean;
}

export const addTestProviders =
  ({
    store = configureStore({ reducer }),
    renderModal = true,
    // @TODO: update tests to use realIntlProvider everywhere
    realIntlProvider = false,
  }: TestProvidersProps): FC<PropsWithChildren<{}>> =>
  ({ children }) => {
    const TranslationsProvider = realIntlProvider ? IntlProvider : I18nProvider;

    return (
      <Provider store={store}>
        <BrowserRouter>
          <ThemeProvider theme={theme}>
            <TranslationsProvider locale="pl" messages={plMessages}>
              <Normalize />
              <GlobalStyles />
              {children}
              {renderModal && <Modal />}
            </TranslationsProvider>
          </ThemeProvider>
        </BrowserRouter>
      </Provider>
    );
  };

const customRender = (
  ui: ReactElement,
  options: Omit<RenderOptions, "options"> = {},
  testProviderOptions: TestProvidersProps = {}
) => render(ui, { wrapper: addTestProviders(testProviderOptions), ...options });

export * from "@testing-library/react";

// override render method
export { customRender as render };
