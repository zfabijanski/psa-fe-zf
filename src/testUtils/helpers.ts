export const sleep = (ms: number = 100): Promise<unknown> =>
  new Promise((resolve) => setTimeout(resolve, ms));
