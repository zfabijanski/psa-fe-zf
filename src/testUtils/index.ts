export * from "./TestProviders";
export * from "./mockCreators";
export * from "./helpers";
