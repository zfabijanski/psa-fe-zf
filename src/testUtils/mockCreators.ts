import {
  ICalculatorCoversConfig,
  IProductCovers,
} from "features/Calculator/types";
import {
  DocumentState,
  IDirectory,
  IDirectoryMetadata,
  IDocumentMetadata,
} from "features/Library/types";
import { IMeeting } from "features/Meeting/History/types";
import { IIllustration, ProductGuid } from "features/Products/types";
import { IModalData, ModalTypes } from "models/Modal";
import {
  Frequency,
  IFundRiskProfileConfig,
  IProductConfig,
  IProductFund,
  IProductGuarantee,
  InvestmentProfile,
  Role,
} from "models/common";
import {
  CalculationStatus,
  Direction,
  ICalculationDetailsDto,
  ICover,
  IInsured,
  YesNo,
} from "models/calculator";

export const createCalculationDetailsDto = (
  overrides?: Partial<ICalculationDetailsDto>
): ICalculationDetailsDto => ({
  calculation_id: 1,
  working_date: "2020-01-01",
  premiums_total: 100,
  guaranteed_death_benefit: 100,
  guaranteed_maturity_benefit: 100,
  projected_maturity_value_egr_m: 100,
  product_categories: [],
  uuid: "uuid",
  version: "version",
  source_system: "source_system",
  product_guid: 1,
  insureds_list: [],
  covers_list: [],
  funds_list: [],
  ...overrides,
});

export const createInsured = (overrides?: Partial<IInsured>): IInsured => ({
  insured_idx: 1,
  insured_uuid: "insured_uuid",
  age_in_months: 1,
  date_of_birth: "2020-01-01",
  role: Role.MAIN,
  ...overrides,
});

export const createCover = (overrides?: Partial<ICover>): ICover => ({
  cover_id: 1,
  cover_idx: 1,
  insured_idx: 1,
  code: "code",
  frequency_code: Frequency.Annually,
  payments: 1,
  direction_of_calculation: Direction.SumAssuredToPremium,
  investment_profile: InvestmentProfile.Balanced,
  indexation: YesNo.Yes,
  sum_assured: 1,
  premium_per_freq: 1,
  duration: 1,
  idx_option: YesNo.Yes,
  rop_option: YesNo.Yes,
  add_sum_assured: 1,
  ending_age: 1,
  exp_mat_ben: 1,
  ...overrides,
});

export const createCalculatorCoversConfig = (
  overrides?: Partial<ICalculatorCoversConfig>
): ICalculatorCoversConfig => ({
  coverName: "coverName",
  isChild: false,
  isHospital: false,
  ...overrides,
});

export const createCalculatorCoversConfigMap = (
  overrides?: Partial<Record<string, ICalculatorCoversConfig>>
): Record<string, ICalculatorCoversConfig> => ({
  coverName1: createCalculatorCoversConfig({ coverName: "coverName1" }),
  coverName2: createCalculatorCoversConfig({ coverName: "coverName2" }),
  ...overrides,
});

export const createProductCovers = (
  overrides?: Partial<IProductCovers>
): IProductCovers => ({
  mainCover: "mainCover",
  additionalCovers: ["additionalCover1", "additionalCover2"],
  wopCovers: ["wopCover1", "wopCover2"],
  childCovers: ["childCover1", "childCover2"],
  ...overrides,
});

export const createProductCoversMap = (
  overrides?: Partial<Record<string, IProductCovers>>
): Record<string, IProductCovers> => ({
  mainCover1: createProductCovers({ mainCover: "mainCover1" }),
  mainCover2: createProductCovers({ mainCover: "mainCover2" }),
  ...overrides,
});

export const createProductConfig = (
  overrides?: Partial<IProductConfig>
): IProductConfig => ({
  product_guid: 1,
  product_name: "product_name",
  product_categories: [],
  product_freqs: [],
  product_covers: [],
  product_funds: [],
  ...overrides,
});

export const createProductConfigMap = (
  overrides?: Partial<Record<string, IProductConfig>>
): Record<string, IProductConfig> => ({
  1: createProductConfig({ product_name: "product_name1" }),
  2: createProductConfig({ product_name: "product_name2" }),
  ...overrides,
});

export const createFundRiskProfileConfig = (
  overrides?: Partial<IFundRiskProfileConfig>
): IFundRiskProfileConfig => ({
  code: "code",
  name_pl: "name_pl",
  name_en: "name_en",
  risk_level: 1,
  status_code: "status_code",
  notes: "notes",
  ...overrides,
});

export const createProductGuarantee = (
  overrides?: Partial<IProductGuarantee>
): IProductGuarantee => ({
  guarantee_code: "guarantee_code",
  guarantee_name: "guarantee_name",
  guarantee_min_value: "1",
  guarantee_duration_months: "4",
  guarantee_percentages: [{ guarantee_percent: 10 }],
  ...overrides,
});

export const createProductFund = (
  overrides?: Partial<IProductFund>
): IProductFund => ({
  cover_code: "cover_code",
  fund_code: "fund_code",
  fund_name: "fund_name",
  fund_risk_profile: "code",
  fund_order: 1,
  guarantees: [createProductGuarantee()],
  ...overrides,
});

export const createModalData = (
  overrides: Partial<IModalData> = {}
): IModalData => ({
  modalIllustration: "alert-fill",
  modalTitleTrKey: "modalTitleTrKey",
  modalContentTrKey: "modalContentTrKey",
  modalContent: "modalContent",
  modalButtons: {
    confirm: { textTrKey: "modal.ok" },
  },
  modalType: ModalTypes.default,
  maskClosable: true,
  ...overrides,
});

export const createIllustration = (
  overrides: Partial<IIllustration> = {}
): IIllustration => ({
  idd_id: 1,
  product_guid: ProductGuid.Savings,
  adequacy_id: 1,
  order_id: 1,
  expected_amount: { value: 3 },
  premium_total: { value: 3 },
  duration: { value: 3 },
  premium_main_cover: { value: 3 },
  frequency: { value: 3 },
  sum_of_premiums: { value: 3 },
  sum_assured: { value: 3 },
  investment_profile: { value: 3 },
  pru_calc_calculation_id: 1,
  status: CalculationStatus.Illustrated,
  json_needs_analysis: null,
  ...overrides,
});

export const createDirectory = (
  overrides?: Partial<IDirectory>
): IDirectory => ({
  directoryId: 1,
  directories: [],
  documentsIds: [1, 2, 3],
  ...overrides,
});

export const createDirectoryMetadata = (
  overrides?: Partial<IDirectoryMetadata>
): IDirectoryMetadata => ({
  directoryId: 1,
  name: "name",
  isOpen: false,
  ...overrides,
});

export const createDocumentMetadata = (
  overrides?: Partial<IDocumentMetadata>
): IDocumentMetadata => ({
  documentId: 1,
  guid: "guid",
  name: "name",
  state: DocumentState.NotAttached,
  isStateChanging: false,
  moduleName: "PRU_LIBRARY",
  ...overrides,
});

export const createMeeting = (overrides?: Partial<IMeeting>): IMeeting => ({
  meetingNo: "meetingNo",
  meetingId: 1,
  name: "name",
  age: 18,
  lastAccessDate: new Date("2023-01-01"),
  createDate: new Date("2023-01-01"),
  statusId: 1,
  statusName: "statusName",
  statusDescription: "statusDescription",
  state: "state",
  haveAdequacy: false,
  haveCalculation: false,
  haveIdd: false,
});
