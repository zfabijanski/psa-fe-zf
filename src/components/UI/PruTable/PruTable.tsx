import { Component } from "react";
import styled from "styled-components";
import { FormattedMessage } from "react-intl";
import { TextEllipsis } from "../../../theme/textEllipsis";
import Row from "./Row";
import SortingArrow from "./SortingArrow";
import { IColumn, IColumnSorting, SortOrder } from "./types";

export const SpanRed = styled.span`
  background-color: ${({ theme }) => theme.newColors.errorBg};
  border-left: 2px solid ${({ theme }) => theme.newColors.error};
  box-shadow: 0px 1px 1px rgba(10, 27, 56, 0.24),
    0px 0px 1px rgba(10, 27, 56, 0.32);
  padding: 0 6px 0 8px;
`;

export const SpanGreen = styled.span`
  background-color: ${({ theme }) => theme.newColors.successBg};
  border-left: 2px solid ${({ theme }) => theme.newColors.success};
  box-shadow: 0px 1px 1px rgba(10, 27, 56, 0.24),
    0px 0px 1px rgba(10, 27, 56, 0.32);
  padding: 0 6px 0 8px;
`;

const ContainerTable = styled.table`
  width: 100%;
  border-collapse: collapse;
  font-size: 16px;
  line-height: 20px;
  table-layout: fixed;
`;

const TrHeader = styled.tr`
  font-size: 14px;
  line-height: 20px;

  user-select: none;
  border: 1px solid ${({ theme }) => theme.newColors.gray40};
`;

const ThHeader = styled.th<
  Pick<IColumn<void>, "width"> & { isActive?: boolean; onClick?: Function }
>`
  white-space: nowrap;
  padding: 15px 12px;
  width: ${({ width }) => width};
  background-color: ${({ theme }) => theme.newColors.white100};
  color: ${({ theme, isActive }) =>
    isActive ? theme.newColors.gray100 : theme.newColors.gray80};
  font-weight: normal;
  position: sticky;
  z-index: 1;
  top: 0;

  ${({ onClick, theme }) =>
    !!onClick &&
    `
    cursor: pointer;

    &:hover {
      color: ${theme.newColors.gray100};
    }
  `};
`;

const Header = styled.div<{ alignRight?: boolean }>`
  display: flex;
  align-items: center;
  justify-content: ${({ alignRight }) => alignRight && "flex-end"};
`;

export interface IPruTableProps<T> {
  primaryKey: keyof T;
  columns: Array<IColumn<T>>;
  data: T[];
  defaultSorting?: IColumnSorting<T>;
  onRowSelect?: (dataRow: T) => void;
  onRowDelete?: (dataRow: T) => void;
  canSelectRow?: (dataRow: T) => boolean;
}

interface IState<T> {
  columns: Array<IColumn<T>>;
  data: T[];
  sorting?: IColumnSorting<T>;
}

export class PruTable<T> extends Component<IPruTableProps<T>, IState<T>> {
  constructor(props: IPruTableProps<T>) {
    super(props);
    this.state = {
      columns: props.columns,
      data: props.data,
    };
  }

  static getDerivedStateFromProps = (
    nextProps: Readonly<IPruTableProps<{}>>,
    nextState: IState<{}>
  ) => {
    const sorting = nextState.sorting || nextProps.defaultSorting;
    if (sorting) {
      const sortingColumn = nextState.columns.find(
        (column) => column.field === sorting.field
      );
      if (sortingColumn) {
        const sorted = nextProps.data.slice().sort(sortingColumn.sorter);
        nextState.data =
          sorting.order === "descend" ? sorted : sorted.reverse();
        nextState.sorting = sorting;
        return nextState;
      }
    } else {
      nextState.data = nextProps.data;
      return nextState;
    }
  };

  private getOrder = (column: IColumn<T>) => {
    const { sorting } = this.state;
    return sorting && sorting.field === column.field
      ? sorting.order
      : undefined;
  };

  private getAlignRight = (column: IColumn<T>) =>
    ["createDate", "lastAccessDate"].includes(column.field.toString());

  private handleSortChange =
    (column: IColumn<T>) => (event?: React.MouseEvent<HTMLElement>) => {
      event?.stopPropagation();

      this.setState((state) => {
        if (state.sorting && state.sorting.field === column.field) {
          return {
            ...state,
            data: state.data.slice().reverse(),
            sorting: {
              field: column.field,
              order: state.sorting.order === "ascend" ? "descend" : "ascend",
            },
          };
        } else {
          return {
            ...state,
            data: state.data.slice().sort(column.sorter),
            sorting: {
              field: column.field,
              order: "descend",
            },
          };
        }
      });
    };

  private handleSelect = (dataRow: T) => () => {
    if (this.props.onRowSelect) {
      this.props.onRowSelect(dataRow);
    }
  };

  private handleDelete = (dataRow: T) => () => {
    if (this.props.onRowDelete) {
      this.props.onRowDelete(dataRow);
    }
  };

  private renderColumnHeader = (column: IColumn<T>, order?: SortOrder) => {
    return (
      <Header alignRight={this.getAlignRight(column)}>
        <TextEllipsis>{column.title}</TextEllipsis>
        {column.sorter && (
          <div>
            <SortingArrow order={order} />
          </div>
        )}
      </Header>
    );
  };

  public render() {
    const { columns, data, sorting } = this.state;
    const { canSelectRow } = this.props;

    const header = (
      <TrHeader>
        {columns.map((column) => (
          <ThHeader
            key={column.field.toString()}
            width={column.width}
            isActive={!!this.getOrder(column)}
            onClick={this.handleSortChange(column)}
            data-testid={`table-header-${column.field.toString()}`}
          >
            {this.renderColumnHeader(column, this.getOrder(column))}
          </ThHeader>
        ))}
        <ThHeader width="86px">
          <Header>
            <FormattedMessage id={"meetingHistory.table.action"} />
          </Header>
        </ThHeader>
      </TrHeader>
    );

    const rows = data.map((dataRow) => (
      <Row
        key={String(dataRow[this.props.primaryKey])}
        primaryKey={this.props.primaryKey}
        dataRow={dataRow}
        columns={columns}
        sorting={sorting}
        onSelect={this.handleSelect(dataRow)}
        onDelete={this.handleDelete(dataRow)}
        isSelectable={canSelectRow === undefined || canSelectRow(dataRow)}
      />
    ));

    return (
      <ContainerTable>
        <thead>{header}</thead>
        <tbody>{rows}</tbody>
      </ContainerTable>
    );
  }
}

export default PruTable;
