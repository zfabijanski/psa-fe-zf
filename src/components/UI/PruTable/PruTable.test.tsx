import { PruTable, IPruTableProps, SpanRed, SpanGreen } from "./PruTable";
import { SortingArrow } from "./SortingArrow";
import { render, screen, fireEvent } from "testUtils";

const mockOnRowSelect = jest.fn();
const mockORowDelete = jest.fn();
const mockCanSelectRow = jest.fn();

const testDataSet: { id: string }[] = [
  { id: "test1" },
  { id: "test2" },
  { id: "test3" },
];

const defaultProps: IPruTableProps<{ id: string }> = {
  primaryKey: "id",
  data: testDataSet,
  columns: [
    {
      field: "id",
      title: "ID",
      width: "100px",
    },
  ],
  onRowSelect: mockOnRowSelect,
  onRowDelete: mockORowDelete,
  canSelectRow: mockCanSelectRow,
};

const setup = (props: Partial<IPruTableProps<{ id: string }>> = {}) =>
  render(<PruTable {...defaultProps} {...props} />);

describe("PruTable", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("should call onRowSelect when row is selected", () => {
    setup();
    fireEvent.click(screen.getByText("test1"));
    expect(mockOnRowSelect).toHaveBeenCalledTimes(1);
  });

  it("should call onRowDelete when row is deleted", () => {
    setup();
    fireEvent.click(screen.getAllByTestId("delete-button")[0]);
    expect(mockORowDelete).toHaveBeenCalledTimes(1);
  });

  it("should call canSelectRow when mounted", () => {
    setup();
    expect(mockCanSelectRow).toHaveBeenCalled();
  });

  it("should allow to onRowSelect to be undefined", () => {
    setup({ onRowSelect: undefined });
    fireEvent.click(screen.getByText("test1"));
    expect(mockOnRowSelect).toHaveBeenCalledTimes(0);
  });

  it("should allow to onRowDelete to be undefined", () => {
    setup({ onRowDelete: undefined });
    fireEvent.click(screen.getAllByTestId("delete-button")[0]);
    expect(mockORowDelete).toHaveBeenCalledTimes(0);
  });

  it("should allow to canSelectRow to be undefined", () => {
    setup({ canSelectRow: undefined });
    fireEvent.click(screen.getByText("test1"));
    expect(mockCanSelectRow).toHaveBeenCalledTimes(0);
  });

  it("should change sort order when clicking on header", () => {
    setup();
    fireEvent.click(screen.getByText("ID"));
    expect(screen.getByTestId("table-header-id")).toHaveStyleRule(
      "color",
      "#2F2F2F"
    );
  });

  it("should change sort order when clicking on header with defaultSorting", () => {
    setup({ defaultSorting: { field: "id", order: "ascend" } });
    fireEvent.click(screen.getByText("ID"));
    expect(screen.getByTestId("table-header-id")).toHaveStyleRule(
      "color",
      "#2F2F2F"
    );
  });

  it('should render custom column with "render" prop', () => {
    setup({
      columns: [
        {
          field: "id",
          title: "ID",
          width: "100px",
          render: (value) => <span>{value.id + "-custom"}</span>,
        },
      ],
    });
    expect(screen.getByText("test1-custom")).toBeInTheDocument();
  });
});

describe("SpanRed", () => {
  it("should render correctly", () => {
    const { container } = render(<SpanRed />);
    expect(container).toMatchSnapshot();
  });
});

describe("SpanGreen", () => {
  it("should render correctly", () => {
    const { container } = render(<SpanGreen />);
    expect(container).toMatchSnapshot();
  });
});

describe("SortingArrow", () => {
  it("should render correctly", () => {
    const { container } = render(<SortingArrow order="descend" />);
    expect(container).toMatchSnapshot();
  });

  it("should render correctly when order is ascend", () => {
    const { container } = render(<SortingArrow order="ascend" />);
    expect(container).toMatchSnapshot();
  });
});
