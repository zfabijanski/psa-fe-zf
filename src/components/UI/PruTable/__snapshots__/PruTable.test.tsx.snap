// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PruTable should render correctly 1`] = `
.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c10 {
  display: inline-block;
  cursor: inherit;
}

.c10 > svg {
  width: 16px;
  height: 100%;
  display: block;
}

.c6 {
  cursor: pointer;
  border: 1px solid #D1D1D1;
}

.c6:hover > * {
  background-color: #F8F8F8;
}

.c7 {
  padding: 13px 24px 13px 12px;
  color: #232424;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c8 {
  text-align: left;
  cursor: pointer;
}

.c9 {
  background-color: #FFFFFF;
  height: 28px;
  width: 28px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 2px solid #FFFFFF;
}

.c9 path {
  stroke-width: 2px;
}

.c9:hover {
  border-color: #055A60;
  background-color: #F9FDFD;
}

.c9:hover path {
  stroke: #055A60;
}

.c9:disabled {
  border-color: #E3E3E3;
  background-color: #EDEDED;
}

.c0 {
  width: 100%;
  border-collapse: collapse;
  font-size: 16px;
  line-height: 20px;
  table-layout: fixed;
}

.c1 {
  font-size: 14px;
  line-height: 20px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid #D1D1D1;
}

.c2 {
  white-space: nowrap;
  padding: 15px 12px;
  width: 100px;
  background-color: #FFFFFF;
  color: #737373;
  font-weight: normal;
  position: -webkit-sticky;
  position: sticky;
  z-index: 1;
  top: 0;
  cursor: pointer;
}

.c2:hover {
  color: #2F2F2F;
}

.c5 {
  white-space: nowrap;
  padding: 15px 12px;
  width: 86px;
  background-color: #FFFFFF;
  color: #737373;
  font-weight: normal;
  position: -webkit-sticky;
  position: sticky;
  z-index: 1;
  top: 0;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <table
    class="c0"
  >
    <thead>
      <tr
        class="c1"
      >
        <th
          class="c2"
          data-testid="table-header-id"
          width="100px"
        >
          <div
            class="c3"
          >
            <div
              class="c4"
            >
              ID
            </div>
          </div>
        </th>
        <th
          class="c5"
          width="86px"
        >
          <div
            class="c3"
          >
            meetingHistory.table.action
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="c6"
      >
        <td
          class="c7"
        >
          test1
        </td>
        <td
          class="c7 c8"
          data-testid="delete-button"
        >
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <svg>
                trash.svg
              </svg>
            </div>
          </div>
        </td>
      </tr>
      <tr
        class="c6"
      >
        <td
          class="c7"
        >
          test2
        </td>
        <td
          class="c7 c8"
          data-testid="delete-button"
        >
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <svg>
                trash.svg
              </svg>
            </div>
          </div>
        </td>
      </tr>
      <tr
        class="c6"
      >
        <td
          class="c7"
        >
          test3
        </td>
        <td
          class="c7 c8"
          data-testid="delete-button"
        >
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              <svg>
                trash.svg
              </svg>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div />
</div>
`;

exports[`SortingArrow should render correctly 1`] = `
.c0 {
  height: 25px;
  width: 25px;
  display: block;
  position: relative;
  cursor: pointer;
}

.c0:before,
.c0:after {
  border: 4px solid transparent;
  content: "";
  display: block;
  height: 0;
  right: 12px;
  top: 50%;
  position: absolute;
  width: 0;
}

.c0:before {
  border-bottom-color: #D1D1D1;
  margin-top: -9px;
}

.c0:after {
  border-top-color: #222;
  margin-top: 1px;
}

.c0:hover {
  opacity: 0.6;
}

<div>
  <span
    class="c0"
    order="descend"
  />
  <div />
</div>
`;

exports[`SortingArrow should render correctly when order is ascend 1`] = `
.c0 {
  height: 25px;
  width: 25px;
  display: block;
  position: relative;
  cursor: pointer;
}

.c0:before,
.c0:after {
  border: 4px solid transparent;
  content: "";
  display: block;
  height: 0;
  right: 12px;
  top: 50%;
  position: absolute;
  width: 0;
}

.c0:before {
  border-bottom-color: #222;
  margin-top: -9px;
}

.c0:after {
  border-top-color: #D1D1D1;
  margin-top: 1px;
}

.c0:hover {
  opacity: 0.6;
}

<div>
  <span
    class="c0"
    order="ascend"
  />
  <div />
</div>
`;

exports[`SpanGreen should render correctly 1`] = `
.c0 {
  background-color: #F0F7EC;
  border-left: 2px solid #7EBD5F;
  box-shadow: 0px 1px 1px rgba(10,27,56,0.24), 0px 0px 1px rgba(10,27,56,0.32);
  padding: 0 6px 0 8px;
}

<div>
  <span
    class="c0"
  />
  <div />
</div>
`;

exports[`SpanRed should render correctly 1`] = `
.c0 {
  background-color: #FBEFEC;
  border-left: 2px solid #D13C15;
  box-shadow: 0px 1px 1px rgba(10,27,56,0.24), 0px 0px 1px rgba(10,27,56,0.32);
  padding: 0 6px 0 8px;
}

<div>
  <span
    class="c0"
  />
  <div />
</div>
`;
